rust/lance/src/dataset/blob.rs

```diff
@@ -33,7 +33,83 @@ pub fn blob_version_from_config(config: &HashMap<String, String>) -> BlobVersion
         .unwrap_or(BlobVersion::V1)
 }
 
-const DEDICATED_THRESHOLD: usize = 4 * 1024 * 1024;
+const INLINE_MAX: usize = 64 * 1024; // 64KB inline cutoff
+const DEDICATED_THRESHOLD: usize = 4 * 1024 * 1024; // 4MB dedicated cutoff
+const PACK_FILE_MAX_SIZE: usize = 1024 * 1024 * 1024; // 1GiB per .pack sidecar
+
+// Maintains rolling `.pack` sidecar files for packed blobs.
+// Layout: data/{data_file_key}/{blob_id:08x}.pack where each file is an
+// unframed concatenation of blob payloads; descriptors store (blob_id,
+// position, size) to locate each slice. A dedicated struct keeps path state
+// and rolling size separate from the per-batch preprocessor logic, so we can
+// reuse the same writer across rows and close/roll files cleanly on finish.
+struct PackWriter {
+    object_store: ObjectStore,
+    data_dir: Path,
+    data_file_key: String,
+    max_pack_size: usize,
+    current_blob_id: Option<u32>,
+    writer: Option<lance_io::object_writer::ObjectWriter>,
+    current_size: usize,
+}
+
+impl PackWriter {
+    fn new(object_store: ObjectStore, data_dir: Path, data_file_key: String) -> Self {
+        Self {
+            object_store,
+            data_dir,
+            data_file_key,
+            max_pack_size: PACK_FILE_MAX_SIZE,
+            current_blob_id: None,
+            writer: None,
+            current_size: 0,
+        }
+    }
+
+    async fn start_new_pack(&mut self, blob_id: u32) -> Result<()> {
+        let path = pack_blob_path(&self.data_dir, &self.data_file_key, blob_id);
+        let writer = self.object_store.create(&path).await?;
+        self.writer = Some(writer);
+        self.current_blob_id = Some(blob_id);
+        self.current_size = 0;
+        Ok(())
+    }
+
+    async fn write_with_allocator<F>(
+        &mut self,
+        alloc_blob_id: &mut F,
+        data: &[u8],
+    ) -> Result<(u32, u64)>
```

Comment:
Can we put at least a `//` comment on this method that explains what is getting returned here?  Looks like the blob_id and position of the data in the pack?